database/migrations/default/1748878240290_non_cr3_ems_trigger/up.sql

```diff
@@ -0,0 +1,232 @@
+                WHERE NEW.case_id = ANY(matched_non_cr3_case_ids)
+                    --
+                    -- If the incident matches any of these conditions, it should NOT be matched
+                    -- to this non-cr3
+                    --
+                    AND (
+                        incident_received_datetime < (NEW.case_timestamp - time_threshold)
+                        OR incident_received_datetime > (NEW.case_timestamp + time_threshold)
+                        OR geometry IS NULL
+                        OR NEW.position IS NULL
+                        OR NEW.is_deleted IS TRUE
+                        OR (NOT ST_DWithin(geometry::geography, NEW.position::geography, meters_threshold))
+                    )
+            )
+            LOOP
+                raise debug 'Handling cleanup for EMS incident # %, ID %', ems_record.incident_number, ems_record.id;
+
+                -- Remove the this non-cr3's case_id from the EMS record's matched_non_cr3_case_ids
+                SELECT array_remove(ems_record.matched_non_cr3_case_ids, NEW.case_id) INTO updated_matched_case_ids;
+                SELECT COALESCE(array_length(updated_matched_case_ids, 1), 0) INTO remaining_case_ids_count;
+
+                raise debug 'remaining_case_ids_count: %', remaining_case_ids_count;
+                raise debug 'updated_matched_case_ids: %', updated_matched_case_ids;
+                --
+                -- Update the EMS records based on remaining match count
+                --
+                IF ems_record.non_cr3_match_status = 'matched_by_manual_qa' THEN
+                    -- For manually matched records, only update the case_id array
+                    raise debug 'Updating matched_non_cr3_case_ids for manually matched EMS incident # %, ID % to %', ems_record.incident_number, ems_record.id, NULLIF(updated_matched_case_ids, '{}');
+                    UPDATE ems__incidents
+                    SET matched_non_cr3_case_ids = NULLIF(updated_matched_case_ids, '{}')
+                    WHERE id = ems_record.id;
+                ELSIF remaining_case_ids_count = 0 THEN
+                    -- No other non-cr3s matched to this EMS record
+                    raise debug 'Updating EMS incident # %, ID % as `umatched`', ems_record.incident_number, ems_record.id;
+                    UPDATE ems__incidents 
+                    SET atd_apd_blueform_case_id = NULL,
+                        non_cr3_match_status = 'unmatched',
+                        matched_non_cr3_case_ids = NULL
+                    WHERE id = ems_record.id;
+                ELSIF remaining_case_ids_count = 1 THEN
+                    -- One non-cr3 remaining: match it to the EMS record
+                    raise debug 'Matching EMS incident # %, ID % to non-cr3 case ID %', ems_record.incident_number, ems_record.id, updated_matched_case_ids[1];
+                    UPDATE ems__incidents 
+                    SET atd_apd_blueform_case_id = updated_matched_case_ids[1],
```

Comment:
if `remaining_case_ids_count = 1` then doesnt that mean there is only one `case_id` for this record, so this should be `updated_matched_case_ids [0]`?

Reply:
'twas a TIL for me that Postgres uses one-based array indexing. `updated_matched_case_ids[1]` grabs the first element!

Reply:
:open_mouth: i never would have guessed that. now im wondering if ive ever indexed incorrectly w Postgres :thinking:

Reply:
+1 - i feel like this is going to be a TIL in the future many times for me 😅 Feels like i should know this already